### PR TITLE
Multithreaded render command encoding

### PR DESCRIPTION
**Pull request by JMS55**
# Objective
- Encoding many GPU commands (such as in a renderpass with many draws, such as the main opaque pass) onto a `wgpu::CommandEncoder` is very expensive, and takes a long time.
- To improve performance, we want to perform the command encoding for these heavy passes in parallel.

## Solution
- `RenderContext` can now queue up "command buffer generation tasks" which are closures that will generate a command buffer when called.
- When finalizing the render context to produce the final list of command buffers, these tasks are run in parallel on the `ComputeTaskPool` to produce their corresponding command buffers.
- The general idea is that the node graph will run in serial, but in a node, instead of doing rendering work, you can add tasks to do render work in parallel with other node's tasks that get ran at the end of the graph execution.

## Nodes Parallelized
- `MainOpaquePass3dNode`
- `PrepassNode`
- `DeferredGBufferPrepassNode`
- `ShadowPassNode` (One task per view)

## Future Work
- For large number of draws calls, might be worth further subdividing passes into 2+ tasks.
- Extend this to UI, 2d, transparent, and transmissive nodes?
- Needs testing - small command buffers are inefficient - it may be worth reverting to the serial command encoder usage for render phases with few items. 
- All "serial" (traditional) rendering work must finish before parallel rendering tasks (the new stuff) can start to run.
- There is still only one submission to the graphics queue at the end of the graph execution. There is still no ability to submit work earlier.

## Performance Improvement
Thanks to @Elabajaba for testing on Bistro.

image

TLDR: Without shadow mapping, this PR has no impact. _With_ shadow mapping, this PR gives **~40 more fps** than main.

---

## Changelog
- `MainOpaquePass3dNode`, `PrepassNode`, `DeferredGBufferPrepassNode`, and each shadow map within `ShadowPassNode` are now encoded in parallel, giving _greatly_ increased CPU performance, mainly when shadow mapping is enabled.
  - Does not work on WASM or AMD+Windows+Vulkan.
- Added `RenderContext::add_command_buffer_generation_task()`.
- `RenderContext::new()` now takes adapter info
- Some render graph and Node related types and methods now have additional lifetime constraints.

## Migration Guide
`RenderContext::new()` now takes adapter info
- Some render graph and Node related types and methods now have additional lifetime constraints.

### CODE REVIEW

**Comment by dmyyy** [comment]
Base implementation looks good to me - interested in seeing what benchmarks will show.

**Comment by JMS55** [comment]
Not sure if I should mark this as a breaking change or not. Technically there's some extra lifetime constraints now, but 99.999% of users won't need to change their code. I think the only thing you would need to change is if you're using bevy_render's RenderContext for your own stuff, independent from the RenderGraph.

**Comment by JMS55** [comment]
Note that this PR currently depends on https://github.com/gfx-rs/wgpu/pull/3626

**Comment by james7132** [comment]
Marking this as blocked on a potential wgpu release, since this otherwise would add complexity without any gains.

**Comment by JMS55** [comment]
Iirc @Elabajaba tried singlethreaded + arcanized and it made ~no difference vs singlethreaded non-arcanized. The bechmarks above were non-arcanized.

**Comment by TheRedXD** [comment]
We've been wanting to use bevy for a fairly graphically intensive game, so any performance improvements we can get would be extremely useful for us! I'd love to see these changes pushed

**Comment by superdump** [comment]
Some clarifications about the benchmarks based on discussion from when this PR was initially made:
* Single-threaded, with shadows, without wgpu arcanisation, and with it were something like 50 vs 54 fps average respectively.
* The image in the original post contains results all with wgpu arcanisation.
* All results so far are based on a much older version of the arcanisation branch.

File: crates/bevy_render/src/renderer/mod.rs [hunk 1]
```diff
@@ -345,25 +346,82 @@ impl RenderContext {
...
+
+        self.command_buffer_queue
+            .push(QueuedCommandBuffer::Ready(command_buffer));
     }
 
-    /// Finalizes the queue and returns the queue of [`CommandBuffer`]s.
+    /// Append a function that will generate a [`CommandBuffer`] to the
+    /// command buffer queue, to be ran later.
+    ///
+    /// If present, this will flush the currently unflushed [`CommandEncoder`]
```
Comment (JMS55) [inline crates/bevy_render/src/renderer/mod.rs:364]: TODO: Awkward wording, need to fix.

**Comment by arcashka** [comment]
Looks fine on 7e837f5bf4e83e2875e012b0dd02051334445992
2024-01-28T09:06:43.311046Z  INFO bevy_render::renderer: AdapterInfo { name: "Intel(R) Graphics (ADL GT2)", vendor: 32902, device: 17958, device_type: IntegratedGpu, driver: "Intel open-source Mesa driver", driver_info: "Mesa 23.0.4-0ubuntu1~23.04.1", backend: Vulkan }
2024-01-28T09:06:43.532619Z  INFO bevy_diagnostic::system_information_diagnostics_plugin::internal: SystemInfo { os: "Linux 23.04 Ubuntu", kernel: "6.2.0-39-generic", cpu: "", core_count: "14", memory: "31.0 GiB" }
image

**Comment by ChristopherBiscardi** [comment]
Looks fine to me on https://github.com/bevyengine/bevy/commit/7e837f5bf4e83e2875e012b0dd02051334445992

```rust
AdapterInfo {
  name: "NVIDIA GeForce RTX 2080 Ti",
  vendor: 4318,
  device: 7687,
  device_type: DiscreteGpu,
  driver: "NVIDIA",
  driver_info: "536.23",
  backend: Vulkan
}

SystemInfo {
  os: "Windows 11 Pro",
  kernel: "22621",
  cpu: "",
  core_count: "16",
  memory: "63.1 GiB"
}
```

image

**Comment by Elabajaba** [comment]
Broken on Windows+AMD for me (shadows flicker):

https://github.com/bevyengine/bevy/assets/177631/c6db1344-acd5-4092-9416-c2e263e06639

```rust
AdapterInfo {
    name: "AMD Radeon RX 6800 XT",
    vendor: 4098,
    device: 29631,
    device_type: DiscreteGpu,
    driver: "AMD proprietary driver",
    driver_info: "24.1.1 (AMD proprietary shader compiler)",
    backend: Vulkan
}
```

**Review by rodolphito — APPROVED** [review]
i can confirm the performance improvements, and additionally there are no non-negligible performance regressions. Less than a tenth of a millisecond slower in very simple scenes that don't have much to parallelize vs several times faster in denser scenes with shadows.

File: crates/bevy_core_pipeline/src/core_3d/main_opaque_pass_3d_node.rs [hunk 2]
```diff
@@ -44,52 +44,69 @@ impl ViewNode for MainOpaquePass3dNode {
...
+            if let Some(viewport) = camera.viewport.as_ref() {
+                render_pass.set_camera_viewport(viewport);
+            }
 
-        // Opaque draws
-        opaque_phase.render(&mut render_pass, world, view_entity);
+            // Opaque draws
+            if !opaque_phase.items.is_empty() {
+                #[cfg(feature = "trace")]
+                let _opaque_main_pass_3d_span = info_span!("opaque_main_pass_3d").entered();
```
Comment (NthTensor) [inline crates/bevy_core_pipeline/src/core_3d/main_opaque_pass_3d_node.rs:80]: Smallest of all nits: I can see that this span has a much narrower scope than the one on line 56, but they have super similar names. Is there a good reason for them to be this similar? I assume it's to maintain compatibility with the previous version?

**Review by NthTensor — APPROVED** [review]
LGTM. Noted performance increase. No flickering shadows.

File: crates/bevy_core_pipeline/src/core_3d/main_opaque_pass_3d_node.rs (hunk 2 above)
Comment (JMS55) [reply crates/bevy_core_pipeline/src/core_3d/main_opaque_pass_3d_node.rs:80]: I think I just copy pasted what the previous version did, yeah

File: crates/bevy_render/src/renderer/mod.rs [hunk 3]
```diff
@@ -345,25 +360,87 @@ impl RenderContext {
...
+        } else {
+            let mut command_buffers = Vec::with_capacity(self.command_buffer_queue.len());
+            let mut task_based_command_buffers = ComputeTaskPool::get().scope(|task_pool| {
+                for (i, queued_command_buffer) in self.command_buffer_queue.into_iter().enumerate()
+                {
+                    match queued_command_buffer {
+                        QueuedCommandBuffer::Ready(command_buffer) => {
+                            command_buffers.push((i, command_buffer));
+                        }
+                        QueuedCommandBuffer::Task(command_buffer_generation_task) => {
```
Comment (james7132) [inline crates/bevy_render/src/renderer/mod.rs:415]: Is there a reason why we can't immediately launch the task on queuing it, and just collect the results in `finish`? As this stands, if we have a large blocking part of the graph, the actual parallel tasks won't get started until that part of the graph is complete.

File: crates/bevy_render/src/renderer/mod.rs [hunk 4]
```diff
@@ -345,25 +360,87 @@ impl RenderContext {
...
+    /// Finalizes and returns the queue of [`CommandBuffer`]s.
+    ///
+    /// This function will wait until all command buffer generation tasks are complete
+    /// by running them in parallel (where supported).
     pub fn finish(mut self) -> Vec<CommandBuffer> {
         self.flush_encoder();
-        self.command_buffers
+
+        if self.command_buffer_queue.is_empty() {
+            self.command_buffer_queue
```
Comment (james7132) [inline crates/bevy_render/src/renderer/mod.rs:399]: This is being done, even though the queue is empty? Shouldn't we just return `Vec::new()`?

**Comment by james7132** [comment]
The changes to the APIs are technically breaking changes. Could you document them in a Migration Guide section?

File: crates/bevy_render/src/renderer/mod.rs (hunk 3 above)
Comment (JMS55) [reply crates/bevy_render/src/renderer/mod.rs:415]: The tasks need to be scoped, which isn't easy to do unless we start them all at the end. For now, this should be fine. I don't foresee it really being a problem.

File: crates/bevy_render/src/renderer/mod.rs (hunk 4 above)
Comment (JMS55) [reply crates/bevy_render/src/renderer/mod.rs]: I don't know why the code was like that. I removed all the redundant stuff.

Comment (NthTensor) [reply crates/bevy_render/src/renderer/mod.rs]: I actually looked at the generated assembly of this, and for whatever reason the original compiled into fewer operations.

**Comment by JMS55** [comment]
It's not a breaking change I don't think, as the lifetimes should be inferred for most use cases.

**Review by james7132 — APPROVED** [review]
Tested this on a few of the stress tests and this does seem to deliver on the performance improvements. `many_foxes` goes from 6.33ms (157 FPS) to 4.68ms (213 FPS).

Overall this looks good to me. Just a few final points that need to be addressed before it can be merged.

**Comment by JMS55** [comment]
> This adds an argument to RenderGraphRunner::run. I don't expect many people to be running their own render graph, but it's worth documenting.
No changes were made to the argument count, just adding some lifetimes that should be inferred for all existing use cases.

File: crates/bevy_pbr/src/render/light.rs [hunk 5]
```diff
@@ -1794,22 +1794,29 @@ impl Node for ShadowPassNode {
                     .get_manual(world, view_light_entity)
                     .unwrap();
 
-                if shadow_phase.items.is_empty() {
-                    continue;
-                }
+                let depth_stencil_attachment =
+                    Some(view_light.depth_attachment.get_attachment(StoreOp::Store));
+
+                render_context.add_command_buffer_generation_task(move |render_device| {
```
[the remark this answers is not in the document]
Comment (JMS55) [reply crates/bevy_pbr/src/render/light.rs:1802]: Added. Technically yes, but I'd rather leave abstraction like that (even if simple) to a future possible render graph API revamp in 0.14.

**Comment by JMS55** [comment]
> Does this still need addressing?

Github won't let me respond to it, but idk what this comment is in reference to.

File: crates/bevy_pbr/src/render/light.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (james7132) [reply crates/bevy_pbr/src/render/light.rs:1802]: OK sounds good to me!

**Comment by james7132** [comment]
Ship it!
